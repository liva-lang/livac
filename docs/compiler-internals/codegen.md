# Code Generation

The code generator emits Rust source code from the IR, handling concurrency transformations, error handling, and runtime integration.

## Location

**File**: `src/codegen.rs` (4,683 lines)

## Overview

The code generator:
1. **Emits Rust Code**: Converts IR → valid Rust syntax
2. **Generates `liva_rt` Module**: Runtime helpers for concurrency and errors
3. **Handles Async/Par**: Transforms concurrency policies into tokio/rayon code
4. **Manages Error Binding**: Implements fallibility system
5. **Tracks Awaitable Tasks**: Lazy await/join optimization

## Architecture

### CodeGenerator

```rust
pub struct CodeGenerator {
    output: String,
    indent_level: usize,
    ctx: DesugarContext,
    in_method: bool,
    in_fallible_function: bool,
    pending_tasks: HashMap<String, TaskInfo>,
    error_binding_vars: HashSet<String>,
    fallible_functions: HashSet<String>,
    // ... other tracking fields
}
```

## Key Responsibilities

### 1. Runtime Module Generation

Generates `liva_rt` module with:

```rust
mod liva_rt {
    use std::future::Future;
    use tokio::task::JoinHandle;
    
    pub struct Error {
        pub message: String,
    }
    
    pub fn spawn_async<F, T>(future: F) -> JoinHandle<T> { ... }
    pub fn fire_async<F>(future: F) { ... }
    pub fn spawn_parallel<F, T>(f: F) -> JoinHandle<T> { ... }
    pub fn fire_parallel<F>(f: F) { ... }
    
    // Display formatting helpers
    pub fn display_array<T: std::fmt::Debug>(arr: &[T]) -> String { ... }
    // ... more helpers
}
```

### 2. Function Generation

**Non-Async, Non-Fallible**:
```rust
fn add(a: i32, b: i32) -> i32 {
    a + b
}
```

**Async**:
```rust
async fn fetchUser(id: i32) -> String {
    let response = reqwest::get(...).await;
    response.text().await.unwrap()
}
```

**Fallible**:
```rust
fn divide(a: i32, b: i32) -> Result<i32, liva_rt::Error> {
    if b == 0 {
        return Err(liva_rt::Error::from("Division by zero"));
    }
    Ok(a / b)
}
```

**Async + Fallible**:
```rust
async fn processData(url: String) -> Result<String, liva_rt::Error> {
    let data = fetch_from_api(&url).await;
    if data.is_empty() {
        return Err(liva_rt::Error::from("No data"));
    }
    Ok(data)
}
```

### 3. Concurrency Transformations

**Async Call (Immediate Await)**:
```liva
let user = async fetchUser(1)
```
↓
```rust
let user = fetchUser(1).await;
```

**Task Async (Deferred Await)**:
```liva
let userTask = task async fetchUser(1)
let user = await userTask
```
↓
```rust
let userTask = liva_rt::spawn_async(async move { fetchUser(1).await });
let user = userTask.await.unwrap();
```

**Parallel Call**:
```liva
let result = par heavyComputation(100)
```
↓
```rust
let result = { heavyComputation(100) };  // Eager evaluation
```

**Task Par**:
```liva
let task = task par compute(100)
let result = await task
```
↓
```rust
let task = liva_rt::spawn_parallel(move || compute(100));
let result = task.await.unwrap();
```

**Fire Async**:
```liva
fire async logEvent("message")
```
↓
```rust
liva_rt::fire_async(async move { logEvent("message").await; });
```

**Fire Par**:
```liva
fire par backgroundCleanup()
```
↓
```rust
liva_rt::fire_parallel(move || { backgroundCleanup(); });
```

### 4. Error Binding

**Basic Error Binding**:
```liva
let result, err = divide(10, 2)
```
↓
```rust
let (result, err) = match divide(10, 2) {
    Ok(val) => (val, "".to_string()),
    Err(e) => (Default::default(), e.to_string()),
};
```

**Error Binding with Async**:
```liva
let result, err = async divide(10, 2)
```
↓
```rust
let (result, err) = match divide(10, 2).await {
    Ok(val) => (val, "".to_string()),
    Err(e) => (Default::default(), e.to_string()),
};
```

**Error Binding with Task**:
```liva
let taskHandle = task async divide(10, 2)
let result, err = await taskHandle
```
↓
```rust
let taskHandle = liva_rt::spawn_async(async move { divide(10, 2).await });
let (result, err) = match taskHandle.await.unwrap() {
    Ok(val) => (val, "".to_string()),
    Err(e) => (Default::default(), e.to_string()),
};
```

### 5. Data-Parallel Loops

**Sequential**:
```liva
for item in items {
    process(item)
}
```
↓
```rust
for item in items {
    process(item);
}
```

**Parallel (Rayon)**:
```liva
for par item in items with threads 4 {
    process(item)
}
```
↓
```rust
items.into_par_iter()
    .with_max_threads(4)
    .for_each(|item| {
        process(item);
    });
```

**Vectorized (SIMD)**:
```liva
for vec value in values with simdWidth 4 {
    compute(value)
}
```
↓
```rust
// SIMD generation (simplified)
for chunk in values.chunks_exact(4) {
    compute(chunk[0]);
    compute(chunk[1]);
    compute(chunk[2]);
    compute(chunk[3]);
}
```

**Parallel + Vec**:
```liva
for parvec value in values with simdWidth 4 ordered {
    process(value)
}
```
↓
```rust
values.into_par_iter()
    .with_simd_width(4)
    .ordered()
    .for_each(|value| {
        process(value);
    });
```

### 6. String Templates

```liva
let name = "Alice"
let greeting = $"Hello, {name}!"
```
↓
```rust
let name = "Alice";
let greeting = format!("Hello, {}!", name);
```

Complex templates:
```liva
let msg = $"User {user.name} (age {user.age}) at {timestamp}"
```
↓
```rust
let msg = format!("User {} (age {}) at {}", user.name, user.age, timestamp);
```

### 7. Classes

```liva
Person {
    constructor(name: string, age: number) {
        this.name = name
        this.age = age
    }
    
    name: string
    age: number
    
    greet() => $"Hello, I'm {this.name}"
}
```
↓
```rust
#[derive(Debug, Clone)]
struct Person {
    name: String,
    age: i32,
}

impl Person {
    fn new(name: String, age: i32) -> Self {
        Self { name, age }
    }
    
    fn greet(&self) -> String {
        format!("Hello, I'm {}", self.name)
    }
}
```

## Main Entry Point

```rust
pub fn generate(module: ir::Module, ctx: DesugarContext) -> Result<String>
```

Returns complete Rust source code ready for compilation.

## Code Organization

### String Building

```rust
fn writeln(&mut self, s: &str)  // Write line with indentation
fn indent(&mut self)            // Increase indent
fn dedent(&mut self)            // Decrease indent
```

### Generation Methods

```rust
fn generate_function(&mut self, func: &ir::Function)
fn generate_statement(&mut self, stmt: &ir::Stmt)
fn generate_expression(&mut self, expr: &ir::Expr)
fn generate_for_loop(&mut self, for_stmt: &ir::For)
fn generate_call(&mut self, call: &ir::CallExpr)
```

## Error Codes

| Code | Description |
|------|-------------|
| **E5001** | Unsupported IR construct |
| **E5002** | Invalid concurrency combination |
| **E5003** | Codegen internal error |

## Output Format

Generated Rust code includes:

1. **Module header**: `use` statements for dependencies
2. **`liva_rt` module**: Runtime helpers
3. **Type definitions**: Structs, enums
4. **Functions**: Public and private
5. **Main function**: Entry point (if present)

## Summary

- **4,683 Lines**: Most complex compiler phase
- **IR → Rust**: Direct code emission
- **Runtime Integration**: Generates `liva_rt` module
- **Concurrency Transformation**: Handles async/par/task/fire
- **Error Binding**: Implements fallibility system
- **Data-Parallel**: Rayon and SIMD code generation

**Next**: [Desugaring →](desugaring.md)

**See Also**:
- [Runtime Module](runtime.md)
- [IR](ir.md)
- [Concurrency](../language-reference/concurrency.md)
