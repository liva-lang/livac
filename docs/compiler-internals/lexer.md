# Lexer (Tokenization)

The lexer converts source code into tokens using the **Logos** crate for high-performance lexical analysis.

## Location

**File**: `src/lexer.rs`

## Overview

The lexer performs:
1. **Tokenization**: Converts character stream to tokens
2. **Comment Stripping**: Removes single-line (`//`) and multi-line (`/* */`) comments
3. **Whitespace Handling**: Skips whitespace automatically
4. **Span Tracking**: Records source locations for error reporting

## Token Types

### Keywords (59 tokens)

**Control Flow**:
- `let`, `const`, `if`, `else`, `while`, `for`, `in`, `switch`, `case`, `default`
- `return`, `throw`, `try`, `catch`

**Concurrency**:
- `async`, `parallel`, `par`, `task`, `fire`, `await`
- `move`, `seq`, `vec`, `parvec`, `with`

**Data-Parallel Policies**:
- `ordered`, `chunk`, `threads`, `simdWidth`
- `prefetch`, `reduction`, `schedule`, `detect`
- `auto`, `safe`, `fast`, `static`, `dynamic`

**Error Handling**:
- `fail`

**Types**:
- `number`, `float`, `bool`, `char`, `string`, `bytes`
- `type`, `import`, `use`, `rust`, `test`

**Literals**:
- `true`, `false`

**Operators** (word-based):
- `and`, `or`, `not`, `as`

### Operators

**Arithmetic**: `+`, `-`, `*`, `/`, `%`

**Comparison**: `==`, `!=`, `<`, `<=`, `>`, `>=`

**Logical**: `&&`, `||`, `!`

**Other**: `=`, `.`, `..`, `=>`, `?`, `:`

### Delimiters

`(`, `)`, `{`, `}`, `[`, `]`, `,`, `;`

### Literals

```rust
IntLiteral(i64)         // 42, 100, 1_000_000
FloatLiteral(f64)       // 3.14, 1.5, 0.001
StringLiteral(String)   // "hello", "world"
CharLiteral(char)       // 'a', 'Z', '\n'
StringTemplate(String)  // $"Hello {name}"
```

### Identifiers

```rust
Ident(String)           // public: myVar, calculate
ProtectedIdent(String)  // protected: _helper
PrivateIdent(String)    // private: __secret
```

Visibility is determined by leading underscores:
- No prefix: Public
- `_`: Protected
- `__`: Private

## Implementation

### Using Logos

```rust
#[derive(Logos, Debug, Clone, PartialEq)]
#[logos(skip r"[ \t\r\n\f]+")]         // Skip whitespace
#[logos(skip r"//[^\n]*")]             // Skip single-line comments
#[logos(skip r"/\*([^*]|\*[^/])*\*/")]  // Skip multi-line comments
pub enum Token {
    #[token("let")]
    Let,
    
    #[token("async")]
    Async,
    
    #[regex(r"[0-9][0-9_]*", parse_int)]
    IntLiteral(i64),
    
    // ... more tokens
}
```

### Key Features

1. **Automatic Whitespace Skipping**: Using `#[logos(skip)]`
2. **Regex Patterns**: For numbers, strings, identifiers
3. **Underscore Separators**: `1_000_000` → `1000000`
4. **String Templates**: `$"Hello {name}"` tokenized separately

### Integer Parsing

```rust
#[regex(r"[0-9][0-9_]*", |lex| {
    lex.slice().replace('_', "").parse::<i64>().ok()
})]
IntLiteral(i64),
```

Supports underscores for readability: `1_000_000`

### Float Parsing

```rust
#[regex(r"[0-9][0-9_]*\.[0-9][0-9_]*", |lex| {
    lex.slice().replace('_', "").parse::<f64>().ok()
})]
FloatLiteral(f64),
```

### String Literals

```rust
#[regex(r#""([^"\\]|\\.)*""#, |lex| {
    let s = lex.slice();
    let content = &s[1..s.len()-1];  // Strip quotes
    Some(content.to_string())
})]
StringLiteral(String),
```

### String Templates

```rust
#[regex(r#"\$"([^"\\]|\\.)*""#, |lex| {
    let s = lex.slice();
    let content = &s[2..s.len()-1];  // Strip $" and "
    Some(content.to_string())
})]
StringTemplate(String),
```

## TokenWithSpan

```rust
pub struct TokenWithSpan {
    pub token: Token,
    pub span: Span,
}
```

Tracks source location for error reporting:
- `span.start`: Start byte offset
- `span.end`: End byte offset
- `span.start_position(map)`: (line, column)

## Public API

```rust
// Tokenize source code
pub fn tokenize(source: &str, file_name: &str) -> Result<Vec<TokenWithSpan>>
```

Returns `Vec<TokenWithSpan>` or `CompilerError` on invalid tokens.

## Error Handling

Lexer errors include:
- **Invalid tokens**: Unrecognized characters
- **Malformed literals**: Unclosed strings, invalid numbers
- **Location info**: Line, column, source snippet

## Summary

- **50+ Keywords**: Including all concurrency and data-parallel policy tokens
- **Logos-based**: High-performance lexing
- **Span Tracking**: For precise error messages
- **Comment Stripping**: Automatic via regex
- **Visibility-Aware**: Identifier prefixes determine visibility

**Next**: [Parser →](parser.md)
