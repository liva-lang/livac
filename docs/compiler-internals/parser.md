# Parser (Syntax Analysis)

The parser converts tokens into an **Abstract Syntax Tree (AST)** using recursive descent parsing.

## Location

**File**: `src/parser.rs` (1,754 lines)

## Overview

The parser performs:
1. **Token Stream Processing**: Consumes tokens from lexer
2. **AST Construction**: Builds typed AST nodes
3. **Syntax Validation**: Checks grammar rules
4. **Error Recovery**: Reports syntax errors with context

## Architecture

### Recursive Descent Parser

```rust
pub struct Parser {
    tokens: Vec<TokenWithSpan>,
    current: usize,
    source_file: String,
    source_code: String,
    source_map: SourceMap,
}
```

**Key Characteristics**:
- **Top-down parsing**: Starts from `Program` → `TopLevel` → ... → `Expr`
- **Predictive**: Looks ahead (`peek()`) to decide parsing path
- **Hand-written**: Not generated from grammar (flexibility for error recovery)

## Parsing Entry Points

### Main Entry

```rust
pub fn parse_program(&mut self) -> Result<Program>
```

Returns `Program { items: Vec<TopLevel> }`

### Top-Level Items

```rust
fn parse_top_level(&mut self) -> Result<TopLevel>
```

Parses:
- `import` declarations
- `use rust` declarations
- `type` definitions
- `class` definitions
- Function declarations
- `test` declarations

## Key Parsing Functions

### Functions

```rust
fn parse_function(&mut self, name: String) -> Result<FunctionDecl>
```

Handles:
- Type parameters: `<T, U>`
- Parameters with types and defaults
- Return type annotations
- Block body or arrow expression body

### Classes

```rust
fn parse_class(&mut self) -> Result<ClassDecl>
```

Features:
- Inheritance: `ChildClass: ParentClass`
- Constructor with parameters
- Fields with visibility
- Methods (block or arrow)

### Statements

```rust
fn parse_statement(&mut self) -> Result<Stmt>
```

Statement types:
- **Variable**: `let x = 10` or `let x, err = fallible()`
  - **Destructuring** (v0.10.2+): `let {name, age} = user` or `let [first, second] = array`
- **Constant**: `const MAX = 100`
- **If**: With optional parentheses, single-statement or block
- **While**: `while condition { ... }`
- **For**: With data-parallel policies (`seq`, `par`, `vec`, `parvec`)
- **Switch**: Case-based branching
- **Try-Catch**: Exception handling
- **Return**: `return value`
- **Throw**: `throw error`
- **Fail**: `fail "message"`
- **Expression**: Any expression as statement

### Variable Bindings and Destructuring (v0.10.2+)

```rust
fn parse_let_bindings(&mut self) -> Result<Vec<VarBinding>>
```

Liva supports multiple binding patterns:

**Simple Binding**:
```liva
let name = "Alice"
let age: number = 25
```

**Fallible Binding** (error tuple):
```liva
let result, err = parseInt("123")
let data, error = HTTP.get("https://api.example.com")
```

**Object Destructuring** (v0.10.2+):
```liva
let {name, age} = user
let {name: userName, age: userAge} = user  // with rename
```

**Array Destructuring** (v0.10.2+):
```liva
let [first, second] = array
let [first, , third] = array  // skip elements
let [head, ...tail] = array   // rest pattern
```

**AST Structure**:

```rust
pub struct VarBinding {
    pub pattern: BindingPattern,
    pub type_ref: Option<TypeRef>,
    pub span: Span,
}

pub enum BindingPattern {
    Identifier(String),
    Object(ObjectPattern),
    Array(ArrayPattern),
}

pub struct ObjectPattern {
    pub fields: Vec<ObjectPatternField>,
}

pub struct ObjectPatternField {
    pub key: String,
    pub binding: String,  // renamed binding or same as key
}

pub struct ArrayPattern {
    pub elements: Vec<Option<String>>,  // None for skipped elements
    pub rest: Option<String>,           // rest pattern variable name
}
```

**Helper Methods**:
- `binding.name() -> Option<&str>`: Returns the simple identifier name, or None for complex patterns
- `binding.is_simple() -> bool`: Returns true if it's a simple identifier binding

### For Loops with Policies

```rust
// Parses: for <policy> var in iterable with <options> { body }
if self.match_token(&Token::For) {
    let mut policy = DataParallelPolicy::Seq;
    
    // Parse policy keyword (seq, par, vec, parvec)
    if self.match_token(&Token::Seq) {
        policy = DataParallelPolicy::Seq;
    } else if self.match_token(&Token::Par) {
        policy = DataParallelPolicy::Par;
    } // ... etc
    
    let var = self.parse_identifier()?;
    self.expect(Token::In)?;
    let iterable = self.parse_expression()?;
    
    // Parse with options
    let options = if self.match_token(&Token::With) {
        self.parse_for_options()?
    } else {
        ForPolicyOptions::default()
    };
    
    // Parse body
    self.expect(Token::LBrace)?;
    let body = self.parse_block_stmt()?;
    self.expect(Token::RBrace)?;
    
    return Ok(Stmt::For(ForStmt { var, iterable, policy, options, body }));
}
```

### Expressions (Operator Precedence)

```rust
fn parse_expression(&mut self) -> Result<Expr>
fn parse_ternary(&mut self) -> Result<Expr>
fn parse_logical_or(&mut self) -> Result<Expr>
fn parse_logical_and(&mut self) -> Result<Expr>
fn parse_equality(&mut self) -> Result<Expr>
fn parse_comparison(&mut self) -> Result<Expr>
fn parse_range(&mut self) -> Result<Expr>
fn parse_additive(&mut self) -> Result<Expr>
fn parse_multiplicative(&mut self) -> Result<Expr>
fn parse_unary(&mut self) -> Result<Expr>
fn parse_postfix(&mut self) -> Result<Expr>
fn parse_primary(&mut self) -> Result<Expr>
```

**Precedence Climbing**: Each function handles one precedence level.

### Call Expressions

```rust
fn parse_postfix(&mut self) -> Result<Expr>
```

Handles:
- **Simple calls**: `func(args)`
- **Concurrency policies**: `async func(args)`, `par func(args)`
- **Task creation**: `task async func(args)`
- **Fire-and-forget**: `fire par func(args)`
- **Member access**: `obj.property`
- **Index access**: `arr[index]`
- **Method calls**: `obj.method(args)`

### Primary Expressions

```rust
fn parse_primary(&mut self) -> Result<Expr>
```

Parses:
- **Literals**: Numbers, strings, booleans, chars
- **Identifiers**: Variable names
- **Parenthesized**: `(expr)`
- **Array literals**: `[1, 2, 3]`
- **Object literals**: `{ name: "Alice", age: 25 }`
- **Struct literals**: `Person { name: "Alice", age: 25 }`
- **String templates**: `$"Hello {name}"`
- **Lambdas**: `(x, y) => x + y` or `move (x) => { ... }`

### Lambda Parsing

```rust
fn parse_lambda_expr(&mut self) -> Result<LambdaExpr>
```

Features:
- **Move capture**: `move (x) => x * 2`
- **Parameter types**: `(x: number, y: number) => x + y`
- **Return types**: `(x: number): number => x * 2`
- **Block bodies**: `(x) => { return x * 2 }`
- **Expression bodies**: `(x) => x * 2`

## Error Handling

### Error Recovery

```rust
fn error(&self, message: String) -> CompilerError
```

Generates rich error messages with:
- **Line and column**: Precise location
- **Source snippet**: Shows problematic code
- **Error code**: E2xxx codes for parser errors
- **Helpful message**: Explains what went wrong

### Example Error

```
[E2010] Expected ';' after statement
  --> main.liva:15:20
   |
15 |     let x = 10
   |               ^ Expected ';' here
```

## AST Structure

### Top-Level Items

```rust
pub enum TopLevel {
    Import(ImportDecl),
    UseRust(UseRustDecl),
    Type(TypeDecl),
    Class(ClassDecl),
    Function(FunctionDecl),
    Test(TestDecl),
}
```

### Statements

```rust
pub enum Stmt {
    VarDecl(VarDecl),
    ConstDecl(ConstDecl),
    Assign(AssignStmt),
    If(IfStmt),
    While(WhileStmt),
    For(ForStmt),
    Switch(SwitchStmt),
    TryCatch(TryCatchStmt),
    Throw(ThrowStmt),
    Fail(FailStmt),
    Return(ReturnStmt),
    Expr(ExprStmt),
    Block(BlockStmt),
}
```

### Expressions

```rust
pub enum Expr {
    Literal(Literal),
    Identifier(String),
    Binary { op: BinOp, left: Box<Expr>, right: Box<Expr> },
    Unary { op: UnOp, operand: Box<Expr> },
    Ternary { condition, then_expr, else_expr },
    Call(CallExpr),
    Member { object: Box<Expr>, property: String },
    Index { object: Box<Expr>, index: Box<Expr> },
    ObjectLiteral(Vec<(String, Expr)>),
    StructLiteral { type_name: String, fields: Vec<(String, Expr)> },
    ArrayLiteral(Vec<Expr>),
    Lambda(LambdaExpr),
    StringTemplate { parts: Vec<StringTemplatePart> },
    Fail(Box<Expr>),
}
```

### Call Expression (with Policies)

```rust
pub struct CallExpr {
    pub callee: Box<Expr>,
    pub args: Vec<Expr>,
    pub exec_policy: ExecPolicy,
}

pub enum ExecPolicy {
    Normal,
    Async,
    Par,
    TaskAsync,
    TaskPar,
    FireAsync,
    FirePar,
}
```

### For Loop Policies

```rust
pub enum DataParallelPolicy {
    Seq,
    Par,
    Vec,
    ParVec,
}

pub struct ForPolicyOptions {
    pub ordered: bool,
    pub chunk: Option<i64>,
    pub threads: Option<ThreadOption>,
    pub simd_width: Option<SimdWidthOption>,
    pub prefetch: Option<i64>,
    pub reduction: Option<ReductionOption>,
    pub schedule: Option<ScheduleOption>,
    pub detect: Option<DetectOption>,
}
```

## Public API

```rust
pub fn parse(tokens: Vec<TokenWithSpan>, file_name: &str, source_code: &str) 
    -> Result<Program>
```

## Summary

- **Recursive Descent**: Hand-written, predictive parser
- **Rich AST**: Fully typed nodes with source spans
- **Policy-Aware**: Parses concurrency and data-parallel annotations
- **Error Recovery**: Detailed error messages with source context
- **1,754 Lines**: Comprehensive grammar support

**Next**: [Semantic Analysis →](semantic.md)

**See Also**:
- [AST Reference](../language-reference/syntax-overview.md)
- [Grammar](grammar.md)
