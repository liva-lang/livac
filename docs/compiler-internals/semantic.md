# Semantic Analysis

The semantic analyzer validates type correctness, infers async functions, detects fallibility, and performs program-wide validation.

## Location

**File**: `src/semantic.rs` (2,067 lines)

## Overview

The semantic analyzer performs **three passes**:

1. **Definition Collection**: Gather types, functions, signatures
2. **Async Inference**: Detect which functions are async (transitive)
3. **Validation**: Type checking, fallibility checks, awaitable validation

## Key Responsibilities

### 1. Async Inference

**Automatically detects** if a function is async:

```rust
// Compiler infers this is async
fetchUser(id: number): string {
  let data = async httpGet($"/users/{id}")  // Contains async call
  return data.body
}
```

**Fixed-point iteration**: Repeats until no new async functions found.

### 2. Fallibility Detection

Tracks functions containing `fail`:

```rust
divide(a: number, b: number): number {
  if b == 0 fail "Division by zero"  // Detected as fallible
  return a / b
}
```

Fallible functions:
- Return `Result<T, liva_rt::Error>` in codegen
- Can use error binding: `let val, err = divide(10, 2)`

### 3. Type Checking

- **Function signatures**: Parameter types, return types
- **Variable types**: Tracked in symbol table
- **Type compatibility**: Ensures assignments are valid

### 4. Awaitable Validation

Enforces rules for `task` and `fire`:

- **Task handles** must be awaited before function returns
- **Fire calls** cannot be awaited
- **Pending tasks** tracked across scopes

### 5. Data-Parallel Policy Validation

Validates `for` loop policies and options:

```rust
// ❌ Invalid: seq cannot use simdWidth
for seq x in items with simdWidth 4 { }  // Error!

// ✅ Valid: parvec can use simdWidth
for parvec x in items with simdWidth 4 { }
```

**Rules enforced**:
- `seq`: No parallel/SIMD options
- `par`: Can use `chunk`, `threads`, `schedule`, `prefetch`, `reduction`, `detect`
- `vec`: Can use `simdWidth`, `ordered`
- `parvec`: Combines `par` and `vec` options

### 6. Error Binding Validation

Validates error binding patterns:

```rust
// ✅ Valid: Fallible function with 2 bindings
let result, err = divide(10, 2)

// ❌ Invalid: Non-fallible with 2 bindings
let result, err = add(10, 2)  // add() doesn't fail

// ✅ Valid: Can ignore error
let result, _ = divide(10, 2)
```

## Architecture

### SemanticAnalyzer

```rust
pub struct SemanticAnalyzer {
    async_functions: HashSet<String>,
    fallible_functions: HashSet<String>,
    types: HashMap<String, TypeInfo>,
    functions: HashMap<String, FunctionSignature>,
    external_modules: HashSet<String>,
    current_scope: Vec<HashMap<String, Option<TypeRef>>>,
    awaitable_scopes: Vec<HashMap<String, AwaitableInfo>>,
    source_file: String,
    source_code: String,
    source_map: Option<SourceMap>,
}
```

### TypeInfo

```rust
struct TypeInfo {
    name: String,
    fields: HashMap<String, (Visibility, TypeRef)>,
    methods: HashMap<String, (Visibility, bool)>, // (visibility, is_async)
}
```

### FunctionSignature

```rust
struct FunctionSignature {
    params: Vec<Option<TypeRef>>,
    return_type: Option<TypeRef>,
    is_async: bool,
    defaults: Vec<bool>,
}
```

## Public API

```rust
pub fn analyze(program: Program, source_file: &str, source_code: &str) 
    -> Result<Program>
```

Returns validated and annotated `Program`.

## Error Codes

| Code | Description |
|------|-------------|
| **E3001** | Async call outside async context |
| **E3002** | Unawaited task handle |
| **E3003** | Awaiting non-awaitable value |
| **E3004** | Fire call cannot be awaited |
| **E3005** | Await in data-parallel loop |
| **E3006** | Invalid policy option combination |
| **E3007** | Invalid chunk size |
| **E3008** | Invalid thread count |
| **E3009** | Invalid SIMD width |
| **E3010** | Error binding mismatch |

## Examples

### Async Inference

**Input AST**:
```liva
fetchData(url: string): string {
  let response = async httpGet(url)
  return response.body
}
```

**After Semantic Analysis**:
- `fetchData.is_async_inferred = true`
- Added to `async_functions` set

### Fallibility Detection

**Input AST**:
```liva
validateAge(age: number) {
  if age < 0 fail "Negative age"
  if age > 150 fail "Too old"
}
```

**After Semantic Analysis**:
- `validateAge.contains_fail = true`
- Added to `fallible_functions` set

### Error Binding Validation

**Valid**:
```liva
let result, err = divide(10, 2)  // ✅ fallible function, 2 bindings
```

**Invalid**:
```liva
let result, err = add(10, 2)  // ❌ E3010: Non-fallible with 2 bindings
```

## Summary

- **2,067 Lines**: Comprehensive semantic validation
- **Three Passes**: Collect → Infer → Validate
- **Async Inference**: Transitive, fixed-point algorithm
- **Fallibility Tracking**: Detects `fail` usage
- **Policy Validation**: Data-parallel correctness
- **Rich Errors**: E3xxx codes with source context

**Next**: [IR (Intermediate Representation) →](ir.md)

**See Also**:
- [Error System](error-system.md)
- [Concurrency](../language-reference/concurrency.md)
- [Error Handling](../language-reference/error-handling.md)
