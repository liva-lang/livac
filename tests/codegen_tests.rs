use insta::assert_snapshot;
use livac::codegen::generate_with_ast;
use livac::lexer::tokenize;
use livac::parser::parse;
use livac::semantic::analyze;

fn compile_and_generate(source: &str) -> String {
    let tokens = tokenize(source).unwrap();
    let program = parse(tokens, source).unwrap();
    let analyzed_program = analyze(program).unwrap();
    let ctx = livac::desugaring::desugar(analyzed_program.clone()).unwrap();
    let (rust_code, _cargo_toml) = generate_with_ast(&analyzed_program, ctx).unwrap();
    rust_code
}

#[test]
fn test_async_main_generation() {
    let source = r#"
main() {
  print("Hello from async main!")
}
"#;

    let rust_code = compile_and_generate(source);
    assert_snapshot!("async_main", rust_code);
}

#[test]
fn test_function_name_generation() {
    let source = r#"
sum(a, b) = a + b
multiply(x, y) = x * y
greet(name) = "Hello " + name

main() {
  let result1 = sum(2, 3)
  let result2 = multiply(4, 5)
  let message = greet("World")
  
  print(result1)
  print(result2)
  print(message)
}
"#;

    let rust_code = compile_and_generate(source);
    assert_snapshot!("function_names", rust_code);
}

#[test]
fn test_return_type_inference() {
    let source = r#"
add(a, b) = a + b
subtract(x, y) = x - y
multiply(n, m) = n * m

main() {
  let sum = add(10, 5)
  let diff = subtract(10, 3)
  let prod = multiply(4, 6)
  
  print(sum)
  print(diff)
  print(prod)
}
"#;

    let rust_code = compile_and_generate(source);
    assert_snapshot!("return_type_inference", rust_code);
}

#[test]
fn test_mixed_function_types() {
    let source = r#"
simpleAdd(a, b) = a + b

complexAdd(x, y) {
  let result = x + y
  return result
}

main() {
  let simple = simpleAdd(1, 2)
  let complex = complexAdd(3, 4)
  
  print(simple)
  print(complex)
}
"#;

    let rust_code = compile_and_generate(source);
    assert_snapshot!("mixed_functions", rust_code);
}

#[test]
fn test_explicit_return_types() {
    let source = r#"
add(a: number, b: number): number = a + b
greet(name: string): string = "Hello " + name
isEven(n: number): bool = n % 2 == 0

main() {
  let sum = add(5, 3)
  let message = greet("Liva")
  let even = isEven(4)
  
  print(sum)
  print(message)
  print(even)
}
"#;

    let rust_code = compile_and_generate(source);
    assert_snapshot!("explicit_return_types", rust_code);
}

#[test]
fn test_comprehensive_codegen() {
    let source = r#"
// Simple expression-bodied functions
add(a, b) = a + b
multiply(x, y) = x * y

// Function with explicit return type
power(base: number, exp: number): number = base * exp

// Block function
complexCalculation(n: number) {
  let doubled = n * 2
  let squared = doubled * doubled
  return squared
}

// Async main function
main() {
  let sum = add(10, 5)
  let product = multiply(3, 4)
  let powered = power(2, 3)
  let complex = complexCalculation(5)
  
  print("Sum: " + sum)
  print("Product: " + product)
  print("Power: " + powered)
  print("Complex: " + complex)
}
"#;

    let rust_code = compile_and_generate(source);
    assert_snapshot!("comprehensive_codegen", rust_code);
}
