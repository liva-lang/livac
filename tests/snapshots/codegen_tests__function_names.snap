---
source: tests/codegen_tests.rs
expression: rust_code
---
fn sum(a: i32, b: i32) -> i32 {
    (a + b)
}

fn multiply(x: i32, y: i32) -> i32 {
    (x * y)
}

fn greet(name: i32) -> i32 {
    ("Hello " + name)
}

fn main() {
    let result1 = sum(2, 3);
    let result2 = multiply(4, 5);
    let message = greet("World");
    println!("{}", result1);;
    println!("{}", result2);;
    println!("{}", message);;
}
